rpc_util.go

```diff
@@ -838,21 +841,27 @@ func recvAndDecompress(p *parser, s recvCompressor, dc Decompressor, maxReceiveM
 			var uncompressedBuf []byte
 			uncompressedBuf, err = dc.Do(compressed.Reader())
 			if err == nil {
+				if size := len(uncompressedBuf); size > maxReceiveMessageSize {
+					out.Free()
+					// TODO: Revisit the error code. Currently keep it consistent with java
+					// implementation.
+					return nil, status.Errorf(codes.ResourceExhausted, "grpc: received message after decompression larger than max (%d vs. %d)", size, maxReceiveMessageSize)
+				}
```

Comment:
This feels like a good amount of duplication now.

Maybe we can move more logic into `decompress` (e.g. pass in `dc`) in order to avoid that?